File: java-receiver/src/main/java/com/wittsfamily/rcode/parsing/RCodeParser.java
package com.wittsfamily.rcode.parsing;

import com.wittsfamily.rcode.RCode;
import com.wittsfamily.rcode.RCodeParameters;

public class RCodeParser {
    private final RCode rcode;
    private RCodeCommandSequence mostRecent;
    private RCodeCommandSlot[] slots;
    private RCodeBigField bigBig;

    public RCodeParser(RCode rcode, RCodeParameters params) {
        mostRecent = null;
        this.rcode = rcode;
        bigBig = new RCodeBigField(params, true);
        slots = new RCodeCommandSlot[params.slotNum];
        for (int i = 0; i < slots.length; i++) {
            slots[i] = new RCodeCommandSlot(rcode, params, bigBig);
        }
    }

    public void parseNext() {
        RCodeCommandSlot targetSlot = null;
        if (bigBig.getLength() == 0) {
            for (int i = 0; i < slots.length; i++) {
                if (!slots[i].isParsed()) {
                    targetSlot = slots[i];
                    break;
                }
            }
            if (targetSlot != null) {
                if (mostRecent != null && !mostRecent.isFullyParsed()) {
                    parse(targetSlot, mostRecent);
                } else {
                    mostRecent = null;
                    for (int i = 0; i < rcode.getChannels().length; i++) {
                        RCodeCommandChannel ch = rcode.getChannels()[i];
                        if (ch.hasCommandSequence() && ch.getCommandSequence().peekFirst() == null) {
                            RCodeCommandSequence seq = ch.getCommandSequence();
                            if (!ch.getInStream().isLocked()) {
                                seq.getInStream().getSequenceIn().openCommandSequence();
                                mostRecent = seq;
                                parse(targetSlot, seq);
                                break;
                            }
                        }
                    }
                }
            }
        }
    }

    private void parse(RCodeCommandSlot slot, RCodeCommandSequence sequence) {
        if (!sequence.getInStream().getSequenceIn().hasNext()) {
            sequence.setFullyParsed(true);
            sequence.releaseInStream();
        } else {
            boolean worked = slot.parseSingleCommand(sequence.getInStream(), sequence);
            if (!worked) {
                sequence.addLast(slot);
                sequence.getInStream().skipSequence();
                sequence.setFullyParsed(true);
                sequence.releaseInStream();
            } else {
                sequence.addLast(slot);
                if (slot.getEnd() == '\n' || !slot.getCommand().continueLocking(sequence.getChannel())) {
                    sequence.setFullyParsed(true);
                    sequence.releaseInStream();
                }
            }
        }
    }
}

File: generic-cpp-receiver/src/test/cpp/RCode/parsing/RCodeParser.cpp
/*
 * RCodeParser.cpp
 *
 *  Created on: 9 Sep 2020
 *      Author: robert
 */

#include "RCodeParser.hpp"

void RCodeParser::eatWhitespace(RCodeInStream *in) {
    char c = in->peek();
    while (in->hasNext() && (c == ' ' || c == '\t' || c == '\r')) {
        in->read();
        c = in->peek();
    }
}

bool RCodeParser::shouldEatWhitespace(RCodeInStream *in) {
    RCodeLookaheadStream *l = in->getLookahead();
    char c = l->read();
    while (c == ' ' || c == '\t' || c == '\r') {
        c = l->read();
    }
    return c != '&' && c != '|';
}

void RCodeParser::parseNext() {
    RCodeCommandSlot *targetSlot = NULL;
    if (bigBig.getLength() == 0) {
        for (int i = 0; i < RCodeParameters::slotNum; i++) {
            if (!slots[i].isParsed()) {
                targetSlot = &slots[i];
                break;
            }
        }
        if (targetSlot != NULL) {
            if (mostRecent != NULL && mostRecent->isActive()
                    && !mostRecent->isFullyParsed()
                    && !mostRecent->hasFailed()) {
                parse(targetSlot, mostRecent);
            } else {
                if (mostRecent != NULL) {
                    mostRecent->unsetFailed();
                    mostRecent->releaseInStream();
                    mostRecent->getChannel()->unlock();
                }
                mostRecent = NULL;
                for (int i = 0; i < rcode->getChannelNumber(); i++) {
                    RCodeCommandChannel *channel = rcode->getChannels()[i];
                    if (channel->canLock() && channel->hasCommandSequence()
                            && !channel->getCommandSequence()->isActive()) {
                        RCodeCommandSequence *seq =
                                channel->getCommandSequence();
                        if (!channel->getInStream()->isLocked()) {
                            channel->lock();
                            seq->getInStream()->getSequenceIn()->openCommandSequence();
                            if (seq->parseFlags()) {
                                if (seq->isFullyParsed()) {
                                    channel->unlock();
                                    seq->setActive();
                                    seq->releaseInStream();
                                } else {
                                    mostRecent = seq;
                                    parse(targetSlot, seq);
                                    seq->setActive();
                                }
                            } else {
                                seq->releaseInStream();
                                channel->unlock();
                                seq->reset();
                            }
                            break;
                        }
                    }
                }
            }
        }
    }
}
void RCodeParser::parse(RCodeCommandSlot *slot,
        RCodeCommandSequence *sequence) {
    if (!sequence->getInStream()->getSequenceIn()->hasNext()) {
        sequence->setFullyParsed(true);
        sequence->releaseInStream();
        sequence->getChannel()->unlock();
    } else {
        bool worked = slot->parseSingleCommand(sequence->getInStream(),
                sequence);
        if (!worked) {
            sequence->addLast(slot);
            sequence->getInStream()->skipSequence();
            sequence->setFailed();
            sequence->setFullyParsed(true);
            sequence->releaseInStream();
            sequence->getChannel()->unlock();
        } else {
            sequence->addLast(slot);
            if (slot->getEnd() == '\n') {
                sequence->setFullyParsed(true);
                sequence->releaseInStream();
                sequence->getChannel()->unlock();
            }
        }
    }
}

void RCodeParserSetupSlots(RCodeParser *parser) {
    for (int i = 0; i < RCodeParameters::slotNum; ++i) {
        parser->slots[i].setup(&parser->bigBig);
    }
}
